### PR TITLE
[osdp_fixes 1/5] Add events support

### PR DESCRIPTION
**Pull request by sidcha**
This PR brings in various updates such as bug fixes, cleanups and new features to subsys/mgmt/osdp in both control panel and peripheral device modes. These changes have been tested with a st-nucleo-g474re board with secure channel enabled.

This PR is a split version of #46449 to improve review-ability (changes remain the same).

### CODE REVIEW

**Review by str4t0m — DISMISSED** [review]
A few comments about the cleanup, didn't check the event delivery and notification stuff yet.

File: subsys/mgmt/osdp/src/osdp_cp.c
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_cp.c:658]: No, return value of 2 was used to indicate a safe point for caller to yield channel to another PD since we are in between commands. Now it is called `OSDP_CP_ERR_CAN_YIELD` and is returned by initialising ret at function entry.

File: subsys/mgmt/osdp/src/osdp_cp.c
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_cp.c:782]: A later patch `mgmt/osdp: cp: Rework CP state management` removes the need for this flag. This hunk should be part of that patch. Put it back here.

File: subsys/mgmt/osdp/src/osdp_pd.c
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:181]: Having var one per line looks too verbose to me. But I've split the `ret = OSDP_PD_ERR_GENERIC` to a separate line.

**Comment by sidcha** [comment]
@str4t0m, thanks for the review. I've addressed all your comments :). Please take another look.

**Comment by sidcha** [comment]
@str4t0m, ping :)

**Comment by carlescufi** [comment]
@str4t0m could you please take another look?

**Comment by str4t0m** [comment]
@sidcha If you drop the last commit that modifies the API the other commits might still be merged in this release.
Would that be in your interest? I'd then suggest to open an issue/issues this PR addresses.

**Comment by sidcha** [comment]
> @sidcha If you drop the last commit that modifies the API the other commits might still be merged in this release.
Would that be in your interest? I'd then suggest to open an issue/issues this PR addresses.

@str4t0m, I don't have an affinity to any particular release; so long as it gets reviewed/merged, I don't mind waiting :).

**Comment by sidcha** [comment]
Hey folks, what is preventing this PR from bring merged? Is having API changes causing this delay, as @str4t0m pointed out? If so can someone advice how/when such changes have to be made? Please note, I'm not change-littering; already the API changes are made in one commit.

There are active users for this module; using it in production in some capacity AFAICT (we also have have independent test confirmations/reviews from @adakus/@r2r0). I have 4 more PRs (each with API changes) to follow this one with; at this rate it is going to take a long time to get them all merged :)

**Comment by erwango** [comment]
@r2r0, @adakus maybe your are the only ones that could provide meaningful review

File: include/zephyr/mgmt/osdp.h
Comment (r2r0) [inline include/zephyr/mgmt/osdp.h:270]: Fields `id`  and `mfgrep` are not present.

File: include/zephyr/mgmt/osdp.h
Comment (r2r0) [inline include/zephyr/mgmt/osdp.h:309]: Is it really needed information about event processing? 
It is just notification
It looks that this return value is not used in the rest of code.

File: subsys/mgmt/osdp/src/osdp_pd.c
Comment (r2r0) [inline subsys/mgmt/osdp/src/osdp_pd.c:366]: Should be called pd->command_callback in case when command is discarded?

File: subsys/mgmt/osdp/src/osdp_pd.c
Comment (r2r0) [inline subsys/mgmt/osdp/src/osdp_pd.c:259]: This `if `is copy&pasted several times.
Could you please consider small refactoring.

File: include/zephyr/mgmt/osdp.h
Comment (r2r0) [inline include/zephyr/mgmt/osdp.h:333]: Consider using pointer to `const`

**Review by r2r0 — CHANGES_REQUESTED** [review]

File: include/zephyr/mgmt/osdp.h
Comment (sidcha) [reply include/zephyr/mgmt/osdp.h:237]: I haven't used `size_t` here to avoid a cast back to int in some comparisons which also needs to be changed. The switch to `size_t` across this module has been on my TODO list for some time now, haven't gotten around to it. Will take this up in a separate PR.

File: include/zephyr/mgmt/osdp.h
Comment (sidcha) [reply include/zephyr/mgmt/osdp.h:305]: The comment is wrong, `pd_info_t` offsets is the LibOSDP way of doing this. Fixed it now.

Here it is the offset (0-indexed) of this PD in kconfig `OSDP_PD_ADDRESS_LIST` (defined [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/osdp/Kconfig.cp#L15)). This offset number is how a PD is referred to between the application and this module. We could refer to the PD by its address (instead of this offset), but it just introduces an extra translation step in the application which can be avoided with this approach.

File: include/zephyr/mgmt/osdp.h
Comment (sidcha) [reply include/zephyr/mgmt/osdp.h:270]: - s/id/type/
- mfgrep is from a future commit, removed it for now.

File: include/zephyr/mgmt/osdp.h
Comment (sidcha) [reply include/zephyr/mgmt/osdp.h:309]: For now, we don't care if the application failed to handle the event. I've just left it there case we need to do some error handling in future.

File: subsys/mgmt/osdp/src/osdp_pd.c
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:366]: Good catch, we can avoid the callback when the command is discarded.

File: subsys/mgmt/osdp/src/osdp_pd.c
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:259]: Sure.

File: include/zephyr/mgmt/osdp.h
Comment (sidcha) [reply include/zephyr/mgmt/osdp.h:333]: Ack.

File: include/zephyr/mgmt/osdp.h
Comment (sidcha) [reply include/zephyr/mgmt/osdp.h:341]: Since I'm not touching this method, will take a separate look at potential const avenues later.

**Review by r2r0 — DISMISSED** [review]

**Review by adakus — DISMISSED** [review]

File: include/zephyr/mgmt/osdp.h
Comment (sidcha) [reply include/zephyr/mgmt/osdp.h:298]: Ack, thanks.

File: include/zephyr/mgmt/osdp.h
Comment (sidcha) [reply include/zephyr/mgmt/osdp.h:302]: Ack, thanks.

**Comment by sidcha** [comment]
@r2r0, @adakus, thanks for your reviews. The last push was just a rebase and typo fixes, please re-approve this PR. 

cc: @erwango (I have 2 +1s now).

**Review by r2r0 — DISMISSED** [review]

**Comment by erwango** [comment]
> @r2r0, @adakus, thanks for your reviews. The last push was just a rebase and typo fixes, please re-approve this PR.
> 
> cc: @erwango (I have 2 +1s now).

@carlescufi How do we proceed on this? Should we consider this is ok, or is there anyone that can take the time to review this ?

**Comment by stephanosio** [comment]
@str4t0m Could you take another look or dismiss your change request if your concerns were addressed?

**Review by adakus — DISMISSED** [review]

**Comment by carlescufi** [comment]
I've dismissed @str4t0m 's review for now.
If this has not been merged in the next week please ping us on Discord, in the #pr-help channel:
https://chat.zephyrproject.org/

File: subsys/mgmt/osdp/src/osdp_pd.c
Comment (gmarull) [inline subsys/mgmt/osdp/src/osdp_pd.c:354]: nitpick: you can use `byteorder.h` API here.

Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:354]: Thanks, will take a look at this in a subsequent PR.

**Review by r2r0 — APPROVED** [review]

**Review by adakus — APPROVED** [review]

File: subsys/mgmt/osdp/src/osdp_pd.c
Comment (gmarull) [reply subsys/mgmt/osdp/src/osdp_pd.c:358]: unresolved

Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:358]: @gmarull, I wasn't planning on taking this up now (as mentioned earlier). There are other places in `subsys/mgmt/` where such a change needs to be made and it doesn't make sense to overload this PR any further.

**Review by gmarull — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
